File: medihelp-user-service/src/main/java/com/medihelp/user/service/FamilyService.java
package com.medihelp.user.service;

import com.medihelp.common.exception.BadRequestException;
import com.medihelp.common.exception.ResourceNotFoundException;
import com.medihelp.user.dto.*;
import com.medihelp.user.entity.FamilyGroup;
import com.medihelp.user.entity.FamilyMember;
import com.medihelp.user.repository.FamilyGroupRepository;
import com.medihelp.user.repository.FamilyMemberRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class FamilyService {

    private final FamilyGroupRepository groupRepository;
    private final FamilyMemberRepository memberRepository;

    @Transactional
    public FamilyGroupResponse createGroup(UUID userId, FamilyGroupRequest request) {
        FamilyGroup group = FamilyGroup.builder()
                .name(request.getName())
                .createdByUserId(userId)
                .build();
        group = groupRepository.save(group);

        // Add creator as OWNER
        FamilyMember owner = FamilyMember.builder()
                .familyGroup(group)
                .userId(userId)
                .role("OWNER")
                .build();
        memberRepository.save(owner);

        return toGroupResponse(group);
    }

    public List<FamilyGroupResponse> getMyGroups(UUID userId) {
        return groupRepository.findByCreatedByUserId(userId).stream()
                .map(this::toGroupResponse)
                .toList();
    }

    @Transactional
    public FamilyMemberResponse addMember(UUID userId, UUID groupId, FamilyMemberRequest request) {
        FamilyGroup group = groupRepository.findById(groupId)
                .orElseThrow(() -> new ResourceNotFoundException("FamilyGroup", "id", groupId));

        if (!group.getCreatedByUserId().equals(userId)) {
            throw new BadRequestException("Only the group owner can add members");
        }

        if (memberRepository.existsByFamilyGroupIdAndUserId(groupId, request.getUserId())) {
            throw new BadRequestException("User is already a member of this group");
        }

        FamilyMember member = FamilyMember.builder()
                .familyGroup(group)
                .userId(request.getUserId())
                .role(request.getRole() != null ? request.getRole() : "VIEWER")
                .build();
        member = memberRepository.save(member);
        return toMemberResponse(member);
    }

    public List<FamilyMemberResponse> getMembers(UUID groupId) {
        return memberRepository.findByFamilyGroupId(groupId).stream()
                .map(this::toMemberResponse)
                .toList();
    }

    @Transactional
    public void removeMember(UUID userId, UUID groupId, UUID memberId) {
        FamilyGroup group = groupRepository.findById(groupId)
                .orElseThrow(() -> new ResourceNotFoundException("FamilyGroup", "id", groupId));

        if (!group.getCreatedByUserId().equals(userId)) {
            throw new BadRequestException("Only the group owner can remove members");
        }

        FamilyMember member = memberRepository.findById(memberId)
                .orElseThrow(() -> new ResourceNotFoundException("FamilyMember", "id", memberId));
        memberRepository.delete(member);
    }

    private FamilyGroupResponse toGroupResponse(FamilyGroup g) {
        return FamilyGroupResponse.builder()
                .id(g.getId())
                .name(g.getName())
                .createdByUserId(g.getCreatedByUserId())
                .createdAt(g.getCreatedAt())
                .build();
    }

    private FamilyMemberResponse toMemberResponse(FamilyMember m) {
        return FamilyMemberResponse.builder()
                .id(m.getId())
                .familyGroupId(m.getFamilyGroup().getId())
                .userId(m.getUserId())
                .role(m.getRole())
                .addedAt(m.getAddedAt())
                .build();
    }
}

File: medihelp-frontend/src/app/core/services/family.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';

export interface FamilyGroup {
  id?: string;
  name: string;
  createdByUserId?: string;
  createdAt?: string;
}

export interface FamilyMember {
  id?: string;
  familyGroupId?: string;
  userId: string;
  role?: string;       // OWNER | CAREGIVER | VIEWER
  addedAt?: string;
}

@Injectable({ providedIn: 'root' })
export class FamilyService {
  private path = '/api/v1/users/me/family';

  constructor(private api: ApiService) {}

  getGroups(): Observable<FamilyGroup[]> {
    return this.api.get<FamilyGroup[]>(`${this.path}/groups`);
  }

  createGroup(group: FamilyGroup): Observable<FamilyGroup> {
    return this.api.post<FamilyGroup>(`${this.path}/groups`, group);
  }

  getMembers(groupId: string): Observable<FamilyMember[]> {
    return this.api.get<FamilyMember[]>(`${this.path}/groups/${groupId}/members`);
  }

  addMember(groupId: string, member: FamilyMember): Observable<FamilyMember> {
    return this.api.post<FamilyMember>(`${this.path}/groups/${groupId}/members`, member);
  }

  removeMember(groupId: string, memberId: string): Observable<void> {
    return this.api.delete<void>(`${this.path}/groups/${groupId}/members/${memberId}`);
  }
}
